nn/transformer.py

```diff
@@ -0,0 +1,237 @@
+               norm_first: bool = False):
+    """
+    :param dim_model: hidden dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param dropout: Dropout value, PyTorch name: dropout
+    :param activation: activation function
+    :param layer_norm_eps: Epsilon value for layer normalization
+    :param norm_first: if ``True`` will perform normalization before other att and ff operations, otherwise after
+    """
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # to be implemented
+    self.multihead_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)  # to be implemented
+
+    self.linear_ff = nn.Linear(dim_ff)
+    self.linear_out = nn.Linear(dim_model)
+
+    self.norm_first = norm_first
+    self.norm_eps = layer_norm_eps
+    self.activation = activation
+    self.dropout = dropout
+
+  def forward(self, tgt: nn.LayerRef, memory: nn.LayerRef) -> nn.LayerRef:
+    """
+    Two possible forward variants of decoder, defined by self.norm_first
+    """
+    x = tgt
+    if self.norm_first:
+      x = x + self._self_attention_block(nn.layer_norm(x, epsilon=self.norm_eps))
+      x = x + self._multi_head_attention_block(nn.layer_norm(x, epsilon=self.norm_eps), memory)
+      x = x + self._feed_forward_block(nn.layer_norm(x, epsilon=self.norm_eps))
+    else:
+      x = nn.layer_norm(x + self._self_attention_block(x), epsilon=self.norm_eps)
+      x = nn.layer_norm(x + self._multi_head_attention_block(x, memory), epsilon=self.norm_eps)
+      x = nn.layer_norm(x + self._feed_forward_block(x), epsilon=self.norm_eps)
+
+    return x
+
+  def _self_attention_block(self, x: nn.LayerRef) -> nn.LayerRef:
+    x = self.self_attn(x, x, x)  # will change with Attention Modules
+    return nn.dropout(x, self.dropout)
+
+  def _multi_head_attention_block(self, x: nn.LayerRef, mem: nn.LayerRef) -> nn.LayerRef:
+    x = self.multihead_attn(x, mem, mem)  # will change with Attention Modules
+    return nn.dropout(x, self.dropout)
+
+  def _feed_forward_block(self, x: nn.LayerRef) -> nn.LayerRef:
+    x = self.linear_ff(x)
+    x = self.activation(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    x = self.linear_out(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    return x
+
+
+class TransformerDecoder(nn.Module):
+  """
+  Defines the full Decoder of the standard transformer
+  """
+  def __init__(self, decoder_layer: Union[TransformerDecoderLayer, Any], num_layers: int,
+               norm: Callable[[nn.LayerRef], nn.LayerRef] = None, layer_norm_eps: float = 1e-5):
+      """
+      :param decoder_layer: Decoder layer to be stacked num_layers times
+      :param num_layers: Number of layers
+      :param norm: normalization function for output layer normalization
```

Comment:
Why do you have that option here for the decoder, but not for the encoder?

Reply:
Hmm what do you mean? norm is also a paramter for encoder.